source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+// For more information on how the RH hash works and in particular how we do
+// deletions, see:
+// http://codecapsule.com/2013/11/17/robin-hood-hashing-backward-shift-deletion/
+
+#include <aws/common/hash_table.h>
+#include <aws/common/common.h>
+#include <aws/common/math.h>
+#include <aws/common/linked_list.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <limits.h>
+
+// Include lookup3.c so we can (potentially) inline it and make use of the mix() macro.
+#include "private/lookup3.c"
+
+static void suppress_unused_lookup3_func_warnings() {
+    // We avoid making changes to lookup3 if we can avoid it, but since it has functions
+    // we're not using, reference them somewhere to suppress the unused function warning.
+    (void)hashword;
+    (void)hashword2;
+    (void)hashlittle;
+    (void)hashbig;
+}
+
+struct hash_table_entry {
+    struct aws_common_hash_element element;
+    uint64_t hash_code; // hash code (0 signals empty)
+};
+
+struct hash_table_state {
+    aws_hash_fn_t hash_fn;
+    aws_equals_fn_t equals_fn;
+    aws_hash_element_destroy_t destroy_fn;
+    struct aws_allocator *alloc;
+
+    size_t size, entry_count;
+    size_t max_load;
+    // We AND a hash value with mask to get the slot index
+    uint64_t  mask;
+    double max_load_factor;
+    // actually variable length
+    struct hash_table_entry slots[1];
+};
+
+static uint64_t hash_for(struct hash_table_state *state, const void *key) {
+    suppress_unused_lookup3_func_warnings();
+
+    uint64_t hash_code = state->hash_fn(key);
+    if (!hash_code) {
+        hash_code = 1;
+    }
+
+    return hash_code;
+}
+
+static int index_for(struct hash_table_state *map, struct hash_table_entry *entry) {
+    return (int)(entry - map->slots);
+}
+
+#if 0
+// Useful debugging code for anyone working on this in the future
+static uint64_t distance(struct hash_table_state *state, int index) {
+    return (index - state->slots[index].hash_code) & state->mask;
+}
+
+void hash_dump(struct aws_common_hash_table *tbl) {
+    struct hash_table_state *state = tbl->pImpl;
+
+    printf("Dumping hash table contents:\n");
+
+    for (int i = 0; i < state->size; i++) {
+        printf("%7d: ", i);
+        struct hash_table_entry *e = &state->slots[i];
+        if (!e->hash_code) {
+            printf("EMPTY\n");
+        } else {
+            printf("k: %p v: %p hash_code: %lld displacement: %lld\n",
+                e->element.key, e->element.value, e->hash_code,
+                (i - e->hash_code) & state->mask);
+        }
+    }
+}
+#endif
+
+#if 0
+// Not currently exposed as an API. Should we have something like this? Useful for benchmarks
+AWS_COMMON_API void aws_common_hash_table_print_stats(struct aws_common_hash_table *table) {
```

Comment:
Even if it was exposed, you don’t need the dll_export in the c file, you can delete AWS_COMMON_API

Reply:
Also it may be better to take a FILE* param